server/src/main/java/com/thoughtworks/go/server/service/ConfigRepositoryInitializer.java

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * @understands initializing config repositories. Loads the configurations from the last checked out modification on server startup.
+ */
+@Service
+public class ConfigRepositoryInitializer implements ConfigChangedListener, PluginChangeListener, Initializer {
+    private PluginManager pluginManager;
+    private final ConfigRepoService configRepoService;
+    private final MaterialRepository materialRepository;
+    private final GoConfigRepoConfigDataSource goConfigRepoConfigDataSource;
+    private GoConfigService goConfigService;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigRepositoryInitializer.class);
+
+    private boolean isConfigLoaded = false;
+    private boolean hasConfigRepoInitializationCompleted = false;
+    private List<String> loadedConfigRepoPlugins = new ArrayList<>();
+
+    @Autowired
+    public ConfigRepositoryInitializer(PluginManager pluginManager, ConfigRepoService configRepoService, MaterialRepository materialRepository, GoConfigRepoConfigDataSource goConfigRepoConfigDataSource, GoConfigService goConfigService) {
+        this.pluginManager = pluginManager;
+        this.configRepoService = configRepoService;
+        this.materialRepository = materialRepository;
+        this.goConfigRepoConfigDataSource = goConfigRepoConfigDataSource;
+        this.goConfigService = goConfigService;
+
+        this.pluginManager.addPluginChangeListener(this);
+    }
+
+    @Override
+    public void initialize() {
+        this.goConfigService.register(this);
+    }
+
+    @Override
+    public void startDaemon() {
+    }
+
+    @Override
+    public void onConfigChange(CruiseConfig newCruiseConfig) {
+        // mark config loaded only once!
+        if (!this.isConfigLoaded) {
+            this.isConfigLoaded = true;
+            this.initializeConfigRepositories();
+        }
+    }
+
+    @Override
+    public void pluginLoaded(GoPluginDescriptor pluginDescriptor) {
+        String pluginId = pluginDescriptor.id();
+        boolean isConfigRepoPlugin = pluginManager.isPluginOfType(CONFIG_REPO_EXTENSION, pluginId);
+
+        if (isConfigRepoPlugin) {
+            this.loadedConfigRepoPlugins.add(pluginId);
+            this.initializeConfigRepositories();
+        }
+    }
+
+    @Override
+    public void pluginUnLoaded(GoPluginDescriptor pluginDescriptor) {
+        //do nothing
+    }
+
+    private void initializeConfigRepositories() {
+        // return if config repositories have already been initialized.
+        if (this.hasConfigRepoInitializationCompleted) {
+            return;
+        }
+
+        // do nothing if the cruise config is not loaded yet..
+        if (!this.isConfigLoaded) {
+            return;
+        }
+
+        List<String> inUseConfigRepoPlugins = this.configRepoService.getConfigRepos().stream().map(RuleAwarePluginProfile::getPluginId).collect(Collectors.toList());
+        if (this.loadedConfigRepoPlugins.containsAll(inUseConfigRepoPlugins)) {
```

Comment:
@GaneshSPatil with this check, in cases where multiple ConfigRepo plugins are used none of the repos will be parsed if a single plugin is missing. Guess this was not intended.

Reply:
@maheshp - yes, you're right. Will change this check to process the config repositories as and when plugins are laoded.